File: competition/ayna/ayna.py
"""
    Created by mbenlioglu on 1/20/2018
"""


def get_score(str1, str2):
    result = 0
    len1, len2 = len(str1), len(str2)
    i = 0 if len1 <= len2 else len1-len2
    while i < len1:
        j, temp_count = 0, 0
        if str1[i] == str2[j]:
            while i < len1 and j < len2 and str1[i] == str2[j]:
                temp_count += 1
                i += 1
                j += 1
            if temp_count > result:
                result = temp_count
        else:
            i += 1
    return result


def get_max_concat(words):
    scores = [[0 for _ in xrange(len(words))] for _ in xrange(len(words))]

    # set scores
    for j in xrange(len(words)):
        for k in xrange(j + 1, len(words)):
            scores[j][k] = get_score(words[j], words[k])

    # get max score
    j, k = 0, 0
    max_score = 0
    for j in xrange(len(words)):
        for k in xrange(j + 1, len(words) - 1):
            scores[j + 1][k + 1] += scores[j][k]
        if scores[j][k + 1] > max_score:
            max_score = scores[j][k + 1]
    return max_score


if __name__ == '__main__':
    loop_cnt = int(raw_input())
    strings = []
    for _ in xrange(loop_cnt):
        strings.append(raw_input())
    strings.sort()
    print get_max_concat(strings)

File: competition/ayna/ayna.cpp
#include <iostream>
#include <string>
#include <algorithm>

// #define DEBUG


unsigned int get_score (const std::string &lhs, const std::string &rhs)
{ // dumb version of KMP string search
	unsigned int lsize = lhs.size(), rsize = rhs.size();
	unsigned int i = 0, temp, score = 0, lpos = lsize;

	lpos = lhs.rfind(rhs[0], lpos - 1);
	while (lpos != -1 && lpos + rsize >= lsize) {
		temp = 0;
		for(int j = lpos; j < lsize; ++j){
			if(lhs[j] == rhs[i++])
				temp++;
			else{
				temp = i = 0;
				break;
			}
		}
		if (temp > score) score = temp;
		lpos = !lpos ? -1 : lhs.rfind(rhs[0], lpos - 1);
	}

#ifdef DEBUG
	std::cout << "(get_score:)  lhs=" << lhs << " rhs=" << rhs << " score=" << score << "\n";
#endif /*DEBUG*/
	return score;
}


int get_max_concat (std::string words[], size_t size)
{
	unsigned int **scores = new unsigned int*[size];
	for (size_t i = 0; i < size; ++i)
		scores[i] = new unsigned int[size]();

	// set scores
	size_t i, j;
	for (i = 0; i < size; ++i)
		for (j = i + 1; j < size; ++j)
			scores[i][j] = get_score(words[i], words[j]);

	// get max score
	unsigned int maxScore = 0;
	for (i = 0; i < size; ++i)
	{
		for (j = i + 1; j < size - 1; ++j)
			scores[i + 1][j + 1] += scores[i][j];
		if (scores[i][j] > maxScore)
			maxScore = scores[i][j];
	}
	return maxScore;
}


int main(int argc, char const *argv[])
{
	size_t loopCnt = 0;
	std::cin >> loopCnt;

	std::string *words = new std::string[loopCnt];
	for (size_t i = 0; i < loopCnt; ++i)
		std::cin >> words[i];

	std::sort(words, words + loopCnt);

	std::cout <<  get_max_concat(words, loopCnt) << "\n";

	return 0;
}